File: DataStructures/StacksAndQueues/Stacks/ArrayStack.py
# A dynamically resizing stack using an array as an underlying data structure

class ArrayStack:

    # Initialize a stack of size 2 and a pointer the head of the stack along with the capacity
    def __init__(self):
        self.stack = [0] * 2
        self.head = 0
        self.capacity = 2
    
    def push(self, item):
        if self.head == self.capacity:
            self.resize(self.capacity * 2)
        self.stack[self.head] = item
        self.head += 1
    
    def pop(self):
        if self.head == 0:
            return None
        item = self.stack[self.head - 1]
        self.head -= 1
        if self.head > 0 and self.head == self.capacity // 4:
            self.resize(self.capacity // 2)
        return item

    def resize(self, new_capacity):
        self.capacity = new_capacity
        copy = [0] * new_capacity
        for i in range(self.head):
            copy[i] = self.stack[i]
        self.stack = copy

    def show_items(self):
        print(self.stack)


def main():
    s = ArrayStack()
    for i in range(20):
        s.push(i)
    s.show_items()
    for i in range(23):
        print(s.pop())
    s.show_items()


main()

File: DataStructures/StacksAndQueues/Stacks/ArrayStack.java
package DataStructures.StacksAndQueues.Stacks;

import java.util.Arrays;

public class ArrayStack<Item> {
    
    // Initialize a stack of any given type, and maintain a record of the size and head
    private Item[] stack = (Item[]) new Object[2];
    private int head = 0;
    private int capacity = 2;

    public void push(Item item) {
        if (head == capacity) { resize(capacity * 2); }
        stack[head++] = item;
    }

    public Item pop() {
        if (head == 0) { return null; }
        Item item = stack[--head];
        if (head > 0 && head == capacity / 4) { resize(capacity / 2); }
        return item;
    }

    private void resize(int newSize) {
        capacity = newSize;
        Item[] copy = (Item[]) new Object[newSize];
        for (int i = 0; i < head; i++) {
            copy[i] = stack[i];
        }
        stack = copy;
    }

    public void showItems() {
        System.out.println(Arrays.toString(stack));
    }

    /* Test Client */
    public static void main(String[] args) {
        ArrayStack<Integer> stack = new ArrayStack<>();
        for (int i = 0; i < 10; i++) {
            stack.push(i);
        }
        stack.showItems();
        for (int j = 0; j < 13; j++) {
            System.out.println(stack.pop());
        }
        stack.showItems();
    }
}
